client.go

```diff
@@ -1137,6 +1137,23 @@ func LinkReceiverSettle(mode ReceiverSettleMode) LinkOption {
 	}
 }
 
+// LinkSourceFilter sets a name and value on the source filter map. For example,
+// LinkSourceFilter("com.microsoft:session-filter", sessionID) would filter messages based on sessionID in
+// Azure Service Bus.
+func LinkSourceFilter(name string, value interface{}) LinkOption {
+	symName := symbol(name)
+	return func(l *link) error {
+		if l.source == nil {
+			l.source = new(source)
+		}
+		if l.source.Filter == nil {
+			l.source.Filter = make(map[symbol]interface{})
+		}
+		l.source.Filter[symName] = value
```

Comment:
According to the spec:
> Every key in the map MUST be of type symbol, every value MUST be either null or of a described type which provides the archetype filter.

This would need to be changed to use `describedType` as the value. I should probably make the `Filter` map be a `map[symbol]*describedType` to enforce this restriction.

I want to avoid exposing `interface{}` to consumers of this library as much as possible, though the limitations of the type system have caused me to use it extensively in the `Message` type.

If there are a finite number of filters, I'd prefer to create an explicit option for each one. If every implementer is going to create their own filter types I think `LinkSourceFilterString(name, value string) LinkOption` is a better solution.

Reply:
In the case of a string value, what would the describedType code be?

Reply:
In this case I believe `describedType.descriptor` should be set to `symbol(name)`.

Relevant section in the spec:
http://docs.oasis-open.org/amqp/core/v1.0/os/amqp-core-types-v1.0-os.html#section-descriptor-values

Reply:
hmm... The following code seems to work for receiver session filtering using a SB queue with required sessions.

```go
func LinkSourceFilterString(name string, value string) LinkOption {
	symName := symbol(name)
	return func(l *link) error {
		if l.source == nil {
			l.source = new(source)
		}
		if l.source.Filter == nil {
			l.source.Filter = make(map[symbol]interface{})
		}
		l.source.Filter[symName] = value
		fmt.Printf("%v - %v", symName, value)
		return nil
	}
}
```

This code doesn't work (doesn't receive any messages for the specified session ID), but seems to correspond to the specification.
```go
func LinkSourceFilterString(name string, value string) LinkOption {
	symName := symbol(name)
	return func(l *link) error {
		if l.source == nil {
			l.source = new(source)
		}
		if l.source.Filter == nil {
			l.source.Filter = make(map[symbol]interface{})
		}
		l.source.Filter[symName] = describedType{
			descriptor: symName,
			value: value,
		}
		fmt.Printf("%v - %v", symName, value)
		return nil
	}
}
```

I'm going to open a thread with the SB team to understand this issue a bit more.

Reply:
That seems odd. The spec seems fairly explicit, but perhaps I've missed something or the map is somehow being encoded wrong. I'll be interested to know what the SB team has to say.

Reply:
Ok. Looks like this is a bug in the SB implementation. They are going to correct the issue.

For the time being, might be good to close this issue. I'll leave that up to you @vcabbage.

It would be nice to be able to extend LinkOptions in the consuming class, but the link struct is not exported. Have you put any thought into this scenario, if so, what are your thoughts on it?

Reply:
I don't mind leaving this open if it's something you'll need eventually.

As far as extending link options, do you mean something like creating a `LinkSessionFilter` option in a library wrapping this one? Or something else?

The former could be accomplished as:
``` go
func LinkSessionFilter(session string) amqp.LinkOption {
	return amqp.LinkSourceFilterString("com.microsoft:session-filter", session)
}
```

Reply:
Yeah, the `LinkSessionFilter` would work the way you describe. It does require getting the original contribution of `LinkSourceFilterString` into the lib. Perhaps, this makes sense for all cases.

I was thinking specifically of gaining access to the link or the source object directly for dealing with an exceptional case, like a vendor defect or similar. Something that you wouldn't want as part of the library, but would be useful for a consumer of the library.

Reply:
I consider `link` and `source` to be internal implementation details subject to change. I'm not particularly keen on the idea of providing hooks to allow violation of the spec, but I also acknowledge that going outside the spec may be the pragmatic solution to some problems. 

I don't want to encourage forking, but I suppose it would be an option.

Another idea would be to add sub-packages with vendor specific options that are allowed special access to these details via an internal package. I'll have to think about this since it would require some restructuring. I take it your immediate concern is being able to set the session filter, correct?

Reply:
Yes, my current concern is about session filter. I'm good with having a short lived fork until the defect is fixed in SB.